working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+  external Object? get target;
+
+  /// Set the current object pointed by the [WeakRef].
+  external void set target (Object? newTarget);
+}
+```
+
+The following classes are added to `dart:ffi` library:
+
+```dart
+/// Any variable which has a static type that is a subtype of a [Finalizable]
+/// is guaranteed to be alive for the full duration of a scope in which it is
+/// defined.
+///
+/// In other words if an object is referenced by such a variable it is
+/// guaranteed to *not* be considered unreachable for the duration of the scope.
+abstract class Finalizable {
+  external factory Finalizable._();
+}
+
+typedef NativeFinalizer = Void Function(Pointer<Void>);
+typedef NativeFinalizerPtr = Pointer<NativeFunction<NativeFinalizer>>
+
+/// [FinalizationRegistry] which will execute its finalizers as early as
+/// possible without waiting for control to return to the event loop.
+///
+/// Will also invoke finalization callbacks when the isolate which created
+/// this finalization registry is shutting down.
+abstract class NativeFinalizationRegistry<F extends Finalizable>
+    extends FinalizationRegistry<F, Pointer> {
+  /// Creates a finalization registry with the given finalization
+  /// callback.
+  ///
+  /// Note: [callback] is expected to be a native function which can be
+  /// executed outside of a Dart isolate. This means that passing an FFI
+  /// trampoline (a function pointer obtained via [Pointer.fromFunction]) is
+  /// not supported for arbitrary Dart functions. This constructor will throw
+  /// if an unsupported [callback] is passed to it.
+  ///
+  /// [callback] might be invoked on an arbitrary thread and not necessary
+  /// on the same thread that created [FinalizationRegistry].
+  external factory FinalizationRegistry(NativeFinalizerPtr callback);
+
+  /// Same as [super.register] but allows to specify an [externalSize] to
+  /// guide GC heuristics.
+  void register(F value,
+                Pointer finalizationToken,
+                {Object? unregisterToken, int externalSize});
+}
+```
+
+`FinalizationRegistry` was directly modeled after its
+[JavaScript counterpart][MDN FinalizationRegistry] and only supports
+asynchronous finalization, while `NativeFinalizationRegistry` is added to
+`dart:ffi` to allow eager synchronous finalization.
+
+Note differences in API between `FinalizationRegistry` and
+`NativeFinalizationRegistry`:
+
+- `NativeFinalizationRegistry` requires objects which are registered with it
+to implement `Finalizable` interface, which serves as a marker instructing
+optimizing compiler to provide stronger liveness guarantees for an object. This
+interface is our solution to the problem of _premature finalization_.
+- `NativeFinalizationRegistry` is constructed with a _native function_ as a
+callback rather than a Dart function. This is done to guarantee that eager
+synchronous execution of a finalization callback is not going to produce any
+side-effects observable from the pure Dart code.
```

Comment:
(Unless you really, really want to? - by calling back into the Dart VM from native code. Or do you actively prevent that?)

Reply:
You can only enter VM when it is in the right state which allows for it. If you try to enter it when it does not, VM will either reject it or crash with a check failure. (Well, there might be some places where we miss appropriate checks - but we should certainly make sure to add them)